### ISSUE TITLE
Fix Float Division by Zero in Tables.py

### ISSUE THREAD

**Issue by C0sm0cats** [Kozea/WeasyPrint#2276]
> Well, I added the following patch to tables.py:
> 
> ```diff
> 
> diff -r 92bd319015e2 -r 179a3a95e21e weasyprint/layout/tables.py
> 
> --- a/weasyprint/layout/tables.py	Wed Aug 19 12:45:48 2020 +0200
> 
> +++ b/weasyprint/layout/tables.py	Wed Aug 19 12:45:59 2020 +0200
> 
> @@ -705,8 +705,12 @@
> 
>          else:
> 
>              added_widths = [
> 
>                  upper_guess[i] - lower_guess[i] for i in range(len(grid))]
> 
> +            added_widths_sum = sum(added_widths)
> 
> +            if added_widths_sum > 0:
> 
>              available_ratio = (
> 
> -                (assignable_width - sum(lower_guess)) / sum(added_widths))
> 
> +                    (assignable_width - sum(lower_guess)) / added_widths_sum)
> 
> +            else:
> 
> +                available_ratio = 0
> 
>              table.column_widths = [
> 
>                  lower_guess[i] + added_widths[i] * available_ratio
> 
>                  for i in range(len(grid))]
> 
> ```
> 
> I assume that there could be numerous reasons, why the added_withs might be zero.. so I didnt track this any further.  If needbe I can provide a working PDF of output that fails otherwise next week. 

 _Originally posted by @JohannesMunk in [#1190](https://github.com/Kozea/WeasyPrint/issues/1190#issuecomment-761563546)_

**Comment by grewn0uille** [comment]
Hello!

Could you please share a HTML/CSS that reproduces the issue?

**Comment by C0sm0cats** [comment]
Hello,

Unfortunately, I no longer have the HTML from the email I used to generate my PDF. However, I encountered a "Float Division by Zero" error, and applying the patch allowed me to successfully retrieve the content in the PDF. Since the patch doesn’t affect other functionalities, I think it would be beneficial to include it in your code, as others might experience the same issue in the future.

Best regards.

**Comment by grewn0uille** [comment]
Adding code to solve a problem that we can’t reproduce (and then can’t add a test in the tests suite) is more like adding dead code, which is something we don’t want to do. As far as we know, this code doesn’t solve the problem, it just hides it. The problem is that we shouldn’t get 0 here.

**Comment by C0sm0cats** [comment]
Hi!

Thanks for your feedback. I understand your concerns about adding code without being able to reproduce the issue. Regardless, I’ve gone ahead and patched the file on my OS to prevent the "Float Division by Zero" error. The addition of the lines:

python

```python
added_widths_sum = sum(added_widths)
if added_widths_sum > 0:
    available_ratio = (assignable_width - sum(lower_guess)) / added_widths_sum
else:
    available_ratio = 0
```
fixed the problem and successfully included the HTML in the PDF.

Since this involves HTML that I receive in an email, I felt it was necessary to patch the file to avoid this issue in the future.

It seems I'm not the only one who encountered this issue, and it might be related to a specific HTML structure. I’m certain that others will likely face this problem in the future as well. If I find a way to reproduce the issue or gather more details, I'll be sure to share that information.

Best regards!

**Comment by liZe** [comment]
> fixed the problem and successfully included the HTML in the PDF.

Thanks for sharing.

> It seems I'm not the only one who encountered this issue, and it might be related to a specific HTML structure. I’m certain that others will likely face this problem in the future as well.

For sure. Unfortunately, each time there’s no way to reproduce.

> If I find a way to reproduce the issue or gather more details, I'll be sure to share that information. 

Good to know, thanks!

If anyone finds a way to reproduce the bug, please share a document to reproduce in this issue.

**Comment by C0sm0cats** [comment]
I understand your concerns about adding code without a clear way to reproduce the issue. However, while we may not currently have a reproducible case, that doesn't mean the existing code is without flaws.

This patch addresses the potential for a "Float Division by Zero" error, ensuring that the application can handle situations where the added_widths might sum to zero. It’s a preventative measure aimed at enhancing stability, rather than merely hiding the issue.

WR

**Comment by quiyip** [comment]
I also stumbled across this bug in a production environment, causing pdf generations to fail and not sent via mail.

I used this single command to apply the patch in my Dockerfile where I have weasyprint installed:
`sed -i 's|available_ratio = (assignable_width - sum(lower_guess)) / sum(added_widths)|available_ratio = (assignable_width - sum(lower_guess)) / sum(added_widths) if sum(added_widths) > 0 else 0|' /.venv/lib/python3.12/site-packages/weasyprint/layout/table.py; \`

This replaces the faulty line in `table.py` which causes the zero division. Change the path if you don't have python packages installed in `/.venv/lib/python3.12/site-packages/weasyprint/layout/table.py; \`.

Hope this is useful to someone in the future.

I have reproduction HTML available, but it contains proprietary information so I'd have to make some adjustments.

**Comment by liZe** [comment]
> I have reproduction HTML available, but it contains proprietary information so I'd have to make some adjustments.

Cool! You can also share it by mail if you prefer.

**Comment by quiyip** [comment]
> > I have reproduction HTML available, but it contains proprietary information so I'd have to make some adjustments.
> 
> Cool! You can also share it by mail if you prefer.

I have sent a mail with a minimum reproduction project to the email address in your GitHub profile. Thank you for helping with this!

**Comment by liZe** [comment]
> I have sent a mail with a minimum reproduction project to the email address in your GitHub profile. Thank you for helping with this!

Thanks a lot! It’s been fixed in [a2e0bbd](https://github.com/Kozea/WeasyPrint/commit/a2e0bbd691673b3514e91a7ed24b188b70ca6238), feedback is welcome!
